# Zoom and auto image zoom fixes
# Works fine on trunk
--- mozilla/content/html/document/src/nsImageDocument.cpp.orig	2007-08-30 15:41:28.000000000 +0300
+++ mozilla/content/html/document/src/nsImageDocument.cpp	2007-08-30 15:45:20.000000000 +0300
@@ -139,20 +139,21 @@ protected:
 
   PRPackedBool                  mResizeImageByDefault;
   PRPackedBool                  mImageIsOverflowing;
   // mImageIsResized is true if the image is currently resized
   PRPackedBool                  mImageIsResized;
   // mShouldResize is true if the image should be resized when it doesn't fit
   // mImageIsResized cannot be true when this is false, but mImageIsResized
   // can be false when this is true
   PRPackedBool                  mShouldResize;
   PRPackedBool                  mFirstResize;
+  float                         mStoredZoom;
 };
 
 NS_IMPL_ADDREF_INHERITED(ImageListener, nsMediaDocumentStreamListener)
 NS_IMPL_RELEASE_INHERITED(ImageListener, nsMediaDocumentStreamListener)
 
 NS_INTERFACE_MAP_BEGIN(ImageListener)
 NS_INTERFACE_MAP_END_INHERITING(nsMediaDocumentStreamListener)
 
 ImageListener::ImageListener(nsImageDocument* aDocument)
   : nsMediaDocumentStreamListener(aDocument)
@@ -246,20 +247,21 @@ ImageListener::OnStopRequest(nsIRequest*
   return nsMediaDocumentStreamListener::OnStopRequest(request, ctxt, status);
 }
 
 
   // NOTE! nsDocument::operator new() zeroes out all members, so don't
   // bother initializing members to 0.
 
 nsImageDocument::nsImageDocument()
 {
 
+  mStoredZoom = 1.0f;
   // NOTE! nsDocument::operator new() zeroes out all members, so don't
   // bother initializing members to 0.
 
 }
 
 nsImageDocument::~nsImageDocument()
 {
 }
 
 NS_IMPL_ADDREF_INHERITED(nsImageDocument, nsMediaDocument)
@@ -339,20 +341,29 @@ nsImageDocument::SetScriptGlobalObject(n
 {
   // If the script global object is changing, we need to unhook our event
   // listeners on the window.
   nsCOMPtr<nsIDOMEventTarget> target;
   if (mScriptGlobalObject &&
       aScriptGlobalObject != mScriptGlobalObject) {
     target = do_QueryInterface(mScriptGlobalObject);
     target->RemoveEventListener(NS_LITERAL_STRING("resize"), this, PR_FALSE);
     target->RemoveEventListener(NS_LITERAL_STRING("keypress"), this,
                                 PR_FALSE);
+    if (mStoredZoom != 1.0) {
+      nsIPresShell *shell = GetPrimaryShell();
+      if (shell) {
+        nsPresContext* context = shell->GetPresContext();
+        if (context)
+          context->SetFullZoom(mStoredZoom, PR_FALSE);
+      }
+      mStoredZoom = 1.0f;
+    }
   }
 
   // Set the script global object on the superclass before doing
   // anything that might require it....
   nsHTMLDocument::SetScriptGlobalObject(aScriptGlobalObject);
 
   if (aScriptGlobalObject) {
     if (!mRootContent) {
       // Create synthetic document
       nsresult rv = CreateSyntheticDocument();
@@ -400,32 +411,49 @@ nsImageDocument::GetImageRequest(imgIReq
   }
 
   *aImageRequest = nsnull;
   return NS_OK;
 }
 
 NS_IMETHODIMP
 nsImageDocument::ShrinkToFit()
 {
   // Keep image content alive while changing the attributes.
+  nsIPresShell *shell = GetPrimaryShell();
+  NS_ENSURE_TRUE(shell, NS_OK);
+  nsPresContext* context = shell->GetPresContext();
+  NS_ENSURE_TRUE(context, NS_OK);
+  float ctxzoom = context->FullZoom();
+  if (ctxzoom != 1.0) {
+    mStoredZoom = ctxzoom;
+    context->SetFullZoom(1.0, PR_FALSE);
+  }
+
   nsCOMPtr<nsIContent> imageContent = mImageContent;
   nsCOMPtr<nsIDOMHTMLImageElement> image = do_QueryInterface(mImageContent);
   image->SetWidth(PR_MAX(1, NSToCoordFloor(GetRatio() * mImageWidth)));
   image->SetHeight(PR_MAX(1, NSToCoordFloor(GetRatio() * mImageHeight)));
   
   imageContent->SetAttr(kNameSpaceID_None, nsGkAtoms::style,
                         NS_LITERAL_STRING("cursor: -moz-zoom-in"), PR_TRUE);
   
   mImageIsResized = PR_TRUE;
   
   UpdateTitleAndCharset();
 
+  NS_ENSURE_TRUE(context, NS_OK);
+  nsIViewManager* vm = context->GetViewManager();
+  NS_ENSURE_TRUE(vm, NS_OK);
+  nsIScrollableView* view;
+  vm->GetRootScrollableView(&view);
+  NS_ENSURE_TRUE(view, NS_OK);
+  view->ScrollTo(0, 0, NS_VMREFRESH_IMMEDIATE);
   return NS_OK;
 }
 
 NS_IMETHODIMP
 nsImageDocument::RestoreImageTo(PRInt32 aX, PRInt32 aY)
 {
   float ratio = GetRatio();
 
   RestoreImage();
   FlushPendingNotifications(Flush_Layout);
@@ -470,21 +498,29 @@ nsImageDocument::RestoreImage()
     imageContent->SetAttr(kNameSpaceID_None, nsGkAtoms::style,
                           NS_LITERAL_STRING("cursor: -moz-zoom-out"), PR_TRUE);
   }
   else {
     imageContent->UnsetAttr(kNameSpaceID_None, nsGkAtoms::style, PR_TRUE);
   }
   
   mImageIsResized = PR_FALSE;
   
   UpdateTitleAndCharset();
+  
+  if (mStoredZoom == 1.0)
+    return NS_OK;
 
+  nsIPresShell *shell = GetPrimaryShell();
+  NS_ENSURE_TRUE(shell, NS_OK);
+  nsPresContext* context = shell->GetPresContext();
+  NS_ENSURE_TRUE(context, NS_OK);
+  context->SetFullZoom(mStoredZoom, PR_FALSE);
   return NS_OK;
 }
 
 NS_IMETHODIMP
 nsImageDocument::ToggleImageSize()
 {
   mShouldResize = PR_TRUE;
   if (mImageIsResized) {
     mShouldResize = PR_FALSE;
     RestoreImage();
--- mozilla/content/html/document/src/nsImageDocument.cpp.orig	2007-09-02 17:08:54.000000000 +0300
+++ mozilla/content/html/document/src/nsImageDocument.cpp	2007-09-02 18:18:57.000000000 +0300
@@ -346,21 +346,21 @@ nsImageDocument::SetScriptGlobalObject(n
       aScriptGlobalObject != mScriptGlobalObject) {
     target = do_QueryInterface(mScriptGlobalObject);
     target->RemoveEventListener(NS_LITERAL_STRING("resize"), this, PR_FALSE);
     target->RemoveEventListener(NS_LITERAL_STRING("keypress"), this,
                                 PR_FALSE);
     if (mStoredZoom != 1.0) {
       nsIPresShell *shell = GetPrimaryShell();
       if (shell) {
         nsPresContext* context = shell->GetPresContext();
         if (context)
-          context->SetFullZoom(mStoredZoom, PR_FALSE);
+          context->SetFullZoom(mStoredZoom);
       }
       mStoredZoom = 1.0f;
     }
   }
 
   // Set the script global object on the superclass before doing
   // anything that might require it....
   nsHTMLDocument::SetScriptGlobalObject(aScriptGlobalObject);
 
   if (aScriptGlobalObject) {
@@ -415,24 +415,24 @@ nsImageDocument::GetImageRequest(imgIReq
 }
 
 NS_IMETHODIMP
 nsImageDocument::ShrinkToFit()
 {
   // Keep image content alive while changing the attributes.
   nsIPresShell *shell = GetPrimaryShell();
   NS_ENSURE_TRUE(shell, NS_OK);
   nsPresContext* context = shell->GetPresContext();
   NS_ENSURE_TRUE(context, NS_OK);
-  float ctxzoom = context->FullZoom();
+  float ctxzoom = context->GetFullZoom();
   if (ctxzoom != 1.0) {
     mStoredZoom = ctxzoom;
-    context->SetFullZoom(1.0, PR_FALSE);
+    context->SetFullZoom(1.0);
   }
 
   nsCOMPtr<nsIContent> imageContent = mImageContent;
   nsCOMPtr<nsIDOMHTMLImageElement> image = do_QueryInterface(mImageContent);
   image->SetWidth(PR_MAX(1, NSToCoordFloor(GetRatio() * mImageWidth)));
   image->SetHeight(PR_MAX(1, NSToCoordFloor(GetRatio() * mImageHeight)));
   
   imageContent->SetAttr(kNameSpaceID_None, nsGkAtoms::style,
                         NS_LITERAL_STRING("cursor: -moz-zoom-in"), PR_TRUE);
   
@@ -506,21 +506,21 @@ nsImageDocument::RestoreImage()
   
   UpdateTitleAndCharset();
   
   if (mStoredZoom == 1.0)
     return NS_OK;
 
   nsIPresShell *shell = GetPrimaryShell();
   NS_ENSURE_TRUE(shell, NS_OK);
   nsPresContext* context = shell->GetPresContext();
   NS_ENSURE_TRUE(context, NS_OK);
-  context->SetFullZoom(mStoredZoom, PR_FALSE);
+  context->SetFullZoom(mStoredZoom);
   return NS_OK;
 }
 
 NS_IMETHODIMP
 nsImageDocument::ToggleImageSize()
 {
   mShouldResize = PR_TRUE;
   if (mImageIsResized) {
     mShouldResize = PR_FALSE;
     RestoreImage();
